LSPS2/README.md

```diff
@@ -0,0 +1,864 @@
+> blockheight for the timeout of the HTLC arriving to the client,
+> but if a block is mined between when the payer sends out the
+> payment and when the client receives it, then the timeout may
+> now be below the normal `min_final_cltv_expiry_delta` of the
+> client.
+> BOLT requires that the payee fail this with
+> `incorrect_or_unknown_payment_details` with a `height` indicating
+> the new current blockheight, but the LSP cannot differentiate
+> between this failure (which would cause the payer to retry with
+> renewed timeouts), versus a deliberate attempt by the payer
+> and client to waste LSP resources (which should cause the LSP to
+> deny future service to this client).
+> By adding a small +2 margin here, we greatly reduce the incidence
+> of such an accident.
+
+#### Forward Compatibility With Post-BOLT11 Payment Schemes
+
+Ultimately, the client MUST provide the information below to
+the payer:
+
+* The client node ID (or any public key whose private key the
+  client knows).
+* The `jit_channel_scid`.
+* The LSP node ID.
+* The `lsp_cltv_expiry_delta`.
+* The expiry for the payment.
+* Whether it can use MPP or not.
+
+However, the client MAY provide this information to the payer
+by means other than BOLT11 invoice, including in a form that
+is not decodable by the payer, but which can be decoded by
+another LN participant.
+
+The only necessary requirement is for the LSP to see an incoming
+HTLC onion that indicates the `jit_channel_scid` as the next hop,
+and as long as the client is able to cause the payer to do this,
+it is would be compatible with this protocol.
+
+> **Non-normative** For example, the client node ID, the
+> `jit_channel_scid`, the LSP node ID, and the
+> `lsp_cltv_expiry_delta` can be encoded as the last hop of
+> a blinded path, for example in a BOLT12 offer or invoice.
+> The payer would not have direct access to the information,
+> but the introduction point of the blinded path would be able
+> to start the decoding of the blinded path.
+
+### 4.  Payment
+
+The client MAY issue the generated invoice to the payer of the
+client, or equivalently inform the payer of the necessary
+information.
+
+The payer can then generate an outgoing payment part to
+deliver the paid amount to the LSP.
+If the client selected "MPP+fixed-invoice" mode, the payer can
+generate multiple outgoing payments parts.
+
+In "MPP+fixed-invoice" mode, the LSP, if it receives a forward
+where the next hop is `jit_channel_scid`:
+
+* SHOULD wait for payment parts of a single payment hash until all
+  parts sum up to at least `payment_size_msat`.
+* MUST NOT fail with `mpp_timeout`, as the final hop is the client
+  and that error is only allowed at the final hop.
+* MUST fail with `unknown_next_peer` if all parts have **NOT**
+  summed up to at least `payment_size_msat` **AND**
+  `opening_fee_params.valid_until` has passed.
+* MAY fail with `unknown_next_peer` if it receives too many payment
+  parts.
```

Comment:
Can you confirm that this spec (including these specific conditions) are implementable in LND with the current available API?
One of the requirements of this spec is to not rely on implementation developers for bringing this spec to reality.

Reply:
Dunno about how the LND hooks forwarding --- for that matter, I do not know if anything other than LDK is actually capable of reducing the value being forwarded.

Reply:
CLN can do this from the LSP side with `htlc_received` hook using a `sendonion` low-level command, and then feed back the received preimage to the `htlc_received` hook response.  The `htlc_received` hook can also receive by checking `extra_fee` explicitly as well.